File: hourai/extensions/validation/rejectors.py
import humanize
import re
from datetime import datetime
from hourai import utils
from .common import Validator, generalize_filter, split_camel_case
from .storage import BanStorage


LOOSE_DELETED_USERNAME_MATCH = re.compile(r'(?i).*Deleted.*User.*')


class NameMatchRejector(Validator):
    """A suspicion level validator that rejects users for username proximity to
    other users already on the server.
    """

    def __init__(self, *, prefix, filter_func,
                 min_match_length=None, subfield=None, member_selector=None):
        self.filter = filter_func
        self.prefix = prefix
        self.subfield = subfield or (lambda m: m.name)
        self.member_selector = member_selector or (lambda m: m.name)
        self.min_match_length = min_match_length

    async def validate_member(self, ctx):
        member_names = {}
        for guild_member in filter(self.filter, ctx.guild.members):
            name = self.member_selector(guild_member) or ''
            member_names.update({
                p: generalize_filter(p) for p in self._split_name(name)
            })
        field_value = self.subfield(ctx.member)
        for filter_name, regex in member_names.items():
            if re.search(regex, field_value):
                ctx.add_rejection_reason(
                    self.prefix + f'Matches: `{filter_name}`')

    def _split_name(self, name):
        split_name = split_camel_case(name)
        if self.min_match_length is not None:
            split_name = (n for n in split_name
                          if len(n) >= self.min_match_length)
        return split_name


class StringFilterRejector(Validator):
    """A general validator that rejects users that have a field that matches
    a set of predefined list of regexes.
    """

    def __init__(self, *, prefix, filters, full_match=False, subfield=None):
        self.prefix = prefix or ''
        self.filters = [(f, re.compile(generalize_filter(f))) for f in filters]
        self.subfield = subfield or (
            lambda ctx: (u.name for u in ctx.usernames))
        if full_match:
            self.match_func = lambda r: r.match
        else:
            self.match_func = lambda r: r.search
        print(self.filters)

    async def validate_member(self, ctx):
        for field_value in self.subfield(ctx):
            for filter_name, regex in self.filters:
                if self.match_func(regex)(field_value):
                    ctx.add_rejection_reason(
                        self.prefix + f'Matches: `{filter_name}`')


class NewAccountRejector(Validator):
    """A suspicion level validator that rejects users that were recently
    created.
    """

    def __init__(self, *, lookback):
        self.lookback = lookback

    async def validate_member(self, ctx):
        if ctx.member.created_at > datetime.utcnow() - self.lookback:
            lookback_naturalized = humanize.naturaltime(self.lookback)
            ctx.add_rejection_reason(
                f"Account created less than {lookback_naturalized}")


class DeletedAccountRejector(Validator):
    """A suspicion level validator that rejects users that are deleted or have
    tell-tale warning signs of faking a deleted account in the past.
    """

    async def validate_member(self, ctx):
        if utils.is_deleted_user(ctx.member):
            ctx.add_rejection_reason(
                "Deleted users cannot be active on Discord. User has been "
                "deleted by Discord of their own accord or for Trust and "
                "Safety reasons, or is faking account deletion.")

        for username in ctx.usernames:
            is_deleted = utils.is_deleted_username(username.name)
            if LOOSE_DELETED_USERNAME_MATCH.match(username.name) and \
               not is_deleted:
                ctx.add_rejection_reason(
                    f'"{username.name}" does not match Discord\'s deletion '
                    f'patterns. User may have attempted to fake account '
                    f"deletion.")
            elif is_deleted and username.discriminator is not None and \
                    username.discriminator < 100:
                ctx.add_rejection_reason(
                    f'"{username.name}#{username.discriminator}" has an '
                    f'unusual discriminator. These are randomly generated. '
                    f'User may have attempted to fake account deletion.')


class NoAvatarRejector(Validator):
    """A suspicion level validator that rejects users without avatars."""

    async def validate_member(self, ctx):
        if ctx.member.avatar is None:
            ctx.add_rejection_reason("User has no avatar.")


class BannedUserRejector(Validator):
    """A malice level validator that rejects users that are banned on other
    servers.
    """

    def __init__(self, *, min_guild_size):
        self.min_guild_size = min_guild_size

    async def validate_member(self, ctx):
        banned = False
        reasons = set()
        guild_ids = [g.id for g in ctx.bot.guilds if self._is_valid_guild(g)]
        bans = await BanStorage(ctx.bot).get_bans(ctx.member.id, guild_ids)
        for ban in bans:
            guild = ctx.bot.get_guild(ban.guild_id)
            assert self._is_valid_guild(guild)
            if ban.reason is not None and ban.reason not in reasons:
                ctx.add_rejection_reason(
                    f"Banned on another server. Reason: `{ban.reason}`.")
                reasons.add(ban.reason)
            banned = True

        if len(reasons) == 0 and banned:
            ctx.add_rejection_reason("Banned on another server.")

    def _is_valid_guild(self, guild):
        return guild is not None and guild.member_count >= self.min_guild_size


class BannedUserRejector(Validator):
    """A malice level validator that rejects users that share characteristics
    with banned users on the server:
     - Exact username matches (ignoring repeated whitespace and casing).
     - Exact avatar matches.
    """

    async def validate_member(self, ctx):
        if not ctx.guild.me.guild_permissions.ban_members:
            return
        # TODO(james7132): Make this use the Ban cache to pull this information
        bans = await ctx.guild.bans()
        self.__check_usernames(ctx, bans)
        self.__check_avatars(ctx, bans)

    def __check_avatars(self, ctx, bans):
        avatar = ctx.member.avatar
        if avatar is None:
            return
        for ban in bans:
            if avatar != ban.user.avatar:
                continue
            ctx.add_rejection_reason(
                f"Exact avatar match with banned user: `{str(ban.user)}`.")

    def __check_usernames(self, ctx, bans):
        normalized_bans = {self._normalize(ban.user.name): ban.user.name
                           for ban in bans}
        for username in ctx.usernames:
            name = self._normalize(username.name)
            for normalized, original in normalized_bans.items():
                if name != normalized:
                    continue
                ctx.add_rejection_reason(
                    f"Exact username match with banned user: `{original}`.")
                break

    def _normalize(self, val):
        return " ".join(val.casefold().split())


class LockdownRejector(Validator):
    """A malice level validator that rejects all users if the guild has been put
    into lockdown.
    """

    async def validate_member(self, ctx):
        guild_state = ctx.bot.guild_states[ctx.guild.id]
        if guild_state.is_locked_down:
            ctx.add_rejection_reason(
                'Lockdown enabled. All new joins must be manually verified.')

File: crates/hourai-logger/src/verification/rejectors.rs
#![allow(clippy::expect_used)]

use super::{context, verifier::*};
use anyhow::Result;
use async_trait::async_trait;
use chrono::Duration;
use chrono::offset::Utc;
use hourai::models::{Snowflake, user::User};
use hourai_sql::{Ban, SqlPool, Username, VerificationBan};
use regex::Regex;

lazy_static! {
    static ref DELETED_USERNAME_MATCH: Regex =
        Regex::new("Deleted User [0-9a-fA-F]{8}").expect("Valid deleted user regex");
    static ref LOOSE_DELETED_USERNAME_MATCH: Regex =
        Regex::new("(?i).*Deleted.*User.*").expect("Valid deleted user regex");
}

fn is_user_deleted(user: &User) -> bool {
    user.avatar.is_none() && DELETED_USERNAME_MATCH.is_match(user.name.as_str())
}

struct DeletedUserRejector(SqlPool);

#[async_trait]
impl Verifier for DeletedUserRejector {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        if is_user_deleted(&ctx.member().user) {
            ctx.add_rejection_reason(
                "Deleted users cannot be active on Discord. User has been \
                deleted by Discord of their own accord or for Trust and \
                Safety reasons, or is faking account deletion.",
            );
        }

        let usernames: Vec<Username> = Username::fetch(ctx.member().user.id, None)
            .fetch_all(&self.0)
            .await?;

        for username in usernames {
            let name = username.name.as_str();
            let is_deleted = DELETED_USERNAME_MATCH.is_match(name);
            if !is_deleted && LOOSE_DELETED_USERNAME_MATCH.is_match(name) {
                ctx.add_rejection_reason(format!(
                    "\"{}\" does not match Discord\'s deletion patterns. User may have \
                             attempted to fake account deletion.",
                    name
                ));
            } else if is_deleted && username.discriminator.map(|d| d < 100).unwrap_or(false) {
                let disc = username.discriminator.unwrap_or(0);
                ctx.add_rejection_reason(format!(
                    "\"{}#{:04}\" has an unusual discriminator for a deleted user. These \
                             are randomly generated. User may have attempted to fake account \
                             deletion.",
                    name, disc
                ));
            }
        }

        Ok(())
    }
}

struct BannedUserRejector {
    sql: SqlPool,
    min_guild_size: u64,
}

#[async_trait]
impl Verifier for BannedUserRejector {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        let bans: Vec<Ban> = Ban::fetch_user_bans(ctx.member().user.id)
            .fetch_all(&self.sql)
            .await?;

        let mut reasons: Vec<Option<String>> = Vec::new();
        for ban in bans {
            let count = hourai_sql::Member::count_guild_members(
                ban.guild_id(),
                /*include_bots=*/ false,
            )
            .fetch_one(&self.sql)
            .await?;
            if count.0 as u64 >= self.min_guild_size {
                reasons.push(ban.reason);
            }
        }

        if !reasons.is_empty() {
            let mut reason = format!("Banned from {} servers", reasons.len());
            let list: String = reasons
                .into_iter()
                .flatten()
                .collect::<Vec<String>>()
                .join("\n");
            if !list.is_empty() {
                reason.push_str(" for the following reasons\n");
                reason.push_str(list.as_str());
            }
            ctx.add_rejection_reason(reason);
        }

        Ok(())
    }
}

struct BannedUsernameRejector(SqlPool);

#[async_trait]
impl Verifier for BannedUsernameRejector {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        let name_bans =
            VerificationBan::fetch_by_name(ctx.guild_id(), ctx.member().user.name.as_str())
                .fetch_all(&self.0)
                .await?;
        for ban in name_bans {
            let mut reason = format!(
                "Exact username match with banned user: {}#{}",
                ban.name, ban.discriminator
            );
            if let Some(ban_reason) = ban.reason {
                reason.push_str(format!(" (Ban Reason: {})", ban_reason).as_str());
            }
            ctx.add_rejection_reason(reason);
        }

        if let Some(ref avatar) = ctx.member().user.avatar {
            let avatar_bans = VerificationBan::fetch_by_avatar(ctx.guild_id(), *avatar)
                .fetch_all(&self.0)
                .await?;
            for ban in avatar_bans {
                let mut reason = format!(
                    "Exact avatar match with banned user: {}#{}",
                    ban.name, ban.discriminator
                );
                if let Some(ban_reason) = ban.reason {
                    reason.push_str(format!(" (Ban Reason: {})", ban_reason).as_str());
                }
                ctx.add_rejection_reason(reason);
            }
        }

        Ok(())
    }
}

pub struct UsernameMatchRejector {
    sql: SqlPool,
    matches: Vec<(String, Regex)>,
    prefix: String,
}

impl UsernameMatchRejector {
    pub fn new(
        sql: SqlPool,
        prefix: impl Into<String>,
        matches: impl IntoIterator<Item = impl Into<String>>,
    ) -> Result<Self> {
        let mut regexes = Vec::new();
        for input in matches {
            let input_str = input.into();
            let regex = Regex::new(&Self::generalize_filter(&input_str))?;
            regexes.push((input_str, regex));
        }
        Ok(Self {
            sql,
            matches: regexes,
            prefix: prefix.into(),
        })
    }

    fn generalize_filter(base: &str) -> String {
        let mut res = String::from("(?i)");
        for ch in base.chars() {
            if ch.is_alphanumeric() {
                res.push(ch);
                res.push('+');
            } else {
                res.push(ch);
            }
        }
        res
    }
}

#[async_trait]
pub trait StringMatchRejector: Send + Sync {
    type Key;
    fn regexes(&self) -> Vec<(Self::Key, Regex)>;
    async fn criteria(&self, ctx: &context::VerificationContext) -> Result<Vec<String>>;
    fn reason(&self, key: &Self::Key, matched: &str) -> String;
}

#[async_trait]
impl StringMatchRejector for UsernameMatchRejector {
    type Key = String;

    fn regexes(&self) -> Vec<(Self::Key, Regex)> {
        self.matches.clone()
    }

    async fn criteria(&self, ctx: &context::VerificationContext) -> Result<Vec<String>> {
        Ok(Username::fetch(ctx.member().user.id, Some(20))
            .fetch_all(&self.sql)
            .await?
            .into_iter()
            .map(|un| un.name)
            .collect())
    }

    fn reason(&self, key: &Self::Key, matched: &str) -> String {
        format!("{}(Matches: {}): {}", self.prefix, key, matched)
    }
}

#[async_trait]
impl<T: StringMatchRejector + Send + Sync> Verifier for T {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        let criteria = self.criteria(ctx).await?;
        let regexes = self.regexes();
        for check in criteria {
            for (key, regex) in &regexes {
                if regex.is_match(check.as_str()) {
                    let reason = self.reason(key, check.as_str());
                    ctx.add_rejection_reason(reason);
                }
            }
        }

        Ok(())
    }
}

pub struct NewAccountRejector(Duration);

#[async_trait]
impl Verifier for NewAccountRejector {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        if ctx.member().created_at() - Utc::now() < self.0 {
            let human_lookback = humantime::format_duration(self.0.to_std().unwrap_or_default());
            ctx.add_rejection_reason(format!("Account created less than {} ago.", human_lookback));
        }
        Ok(())
    }
}

pub fn new_account(lookback: Duration) -> BoxedVerifier {
    Box::new(NewAccountRejector(lookback))
}

pub fn no_avatar() -> BoxedVerifier {
    Box::new(NoAvatarRejector)
}

pub struct NoAvatarRejector;

#[async_trait]
impl Verifier for NoAvatarRejector {
    async fn verify(&self, ctx: &mut context::VerificationContext) -> Result<()> {
        if ctx.member().user.avatar.is_none() {
            ctx.add_rejection_reason("User has no avatar.");
        }
        Ok(())
    }
}

pub fn banned_user(sql: SqlPool, min_guild_size: u64) -> BoxedVerifier {
    Box::new(BannedUserRejector {
        sql,
        min_guild_size,
    })
}

pub fn banned_username(sql: SqlPool) -> BoxedVerifier {
    Box::new(BannedUsernameRejector(sql))
}

pub fn deleted_user(sql: SqlPool) -> BoxedVerifier {
    Box::new(DeletedUserRejector(sql))
}

#[allow(clippy::expect_used)]
pub fn username_match(
    sql: SqlPool,
    prefix: impl Into<String>,
    filters: impl IntoIterator<Item = impl Into<String>>,
) -> BoxedVerifier {
    Box::new(UsernameMatchRejector::new(sql, prefix, filters).expect("valid regex patterns"))
}
